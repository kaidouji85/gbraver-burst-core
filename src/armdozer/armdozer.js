// @flow

/** アームドーザID */
export type ArmDozerId = string;

// TODO バーストの種類を追加する

/** アームドーザ基本情報 */
export type Armdozer = {
  /** アームドーザID */
  id: ArmDozerId,
  /** 名前 */
  name: string,
  /** 最大HP */
  maxHp: number,
  /** 最大バッテリー */
  maxBattery: number,
  /** 攻撃 */
  power: number,
  /** スピード */
  speed: number,
  /** アームドーザの外見を指定する */
  appearance: string,
};
