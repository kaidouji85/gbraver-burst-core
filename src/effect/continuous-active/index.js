// @flow

import type {GameState, GameStateX, TurnChange} from "../..";
import {removeContinuousActive} from "./remove-continuous-active";

/**
 * アクティブプレイヤー継続を実行できるか否かを判定する
 *
 * @param state ゲーム ステート
 * @return 判定結果、trueで実行できる
 */
export function canContinuousActive(state: GameState): boolean {
  const activePlayer = state.players.find(v => v.playerId === state.activePlayerId);
  if (!activePlayer) {
    return false;
  }

  return activePlayer.armdozer.effects
    .filter(v => v.type === 'ContinuousActivePlayer')
    .length > 0;
}


/**
 * アクティブプレイヤー継続を実行する
 *
 * @param state 更新前のゲーム ステート
 * @return 更新結果、実行不可能な場合はnullを返す
 */
export function continuousActive(state: GameState): ?GameStateX<TurnChange> {
  const activePlayer = state.players.find(v => v.playerId === state.activePlayerId);
  if (!activePlayer) {
    return null;
  }

  const updatedActivePlayer = {
    ...activePlayer,
    armdozer: {
      ...activePlayer.armdozer,
      effects: removeContinuousActive(activePlayer.armdozer.effects)
    }
  };
  const updatedPlayers = state.players
    .map(v => v.playerId === activePlayer.playerId ? updatedActivePlayer : v);
  const effect = {
    name: 'TurnChange',
    recoverBattery: 0
  };
  return {
    ...state,
    players: updatedPlayers,
    effect: effect
  };
}