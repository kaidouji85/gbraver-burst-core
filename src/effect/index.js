// @flow

import type {InputCommand} from "./input-command/input-command";
import type {Battle} from "./battle/battle";
import type {TurnChange} from "./turn-change/turn-change";
import type {StartGame} from "./start-game/start-game";
import type {BurstEffect} from "./burst/burst-effect";
import type {GameEnd} from "./game-end/game-end";
import type {BatteryDeclaration} from "./battery-declaration/battery-declaration";
import type {Reflect} from "./reflect/reflect";
import type {UpdateRemainingTurn} from "./update-remaning-turn/update-remaining-turn";

/** ゲーム中に発生する効果をまとめたもの */
export type Effect =
  StartGame |
  GameEnd |
  InputCommand |
  Reflect |
  BatteryDeclaration |
  Battle |
  TurnChange |
  BurstEffect |
  UpdateRemainingTurn;
