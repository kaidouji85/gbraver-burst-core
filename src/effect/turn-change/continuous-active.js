// @flow

import type {ArmdozerEffect, PlayerState} from "../..";

/**
 * 指定したプレイヤーがアクティブプレイヤー継続を持っているか否かを判定する
 *
 * @param player 判定対象
 * @return 判定結果、trueでアクティブプレイヤー継続を持っている
 */
export function hasContinuousActivePlayer(player: PlayerState): boolean {
  return player.armdozer.effects
    .filter(v => v.type === 'ContinuousActivePlayer')
    .length > 0;
}

/**
 * アクティブプレイヤー継続を取り除く
 *
 * @param origin 処理対象
 * @return 処理結果
 */
export function removeContinuousActive(origin: ArmdozerEffect[]): ArmdozerEffect[] {
  const removeTarget = origin.find(v => v.type === 'ContinuousActivePlayer');
  if (!removeTarget) {
    return origin;
  }

  return origin.filter(v => v !== removeTarget);
}