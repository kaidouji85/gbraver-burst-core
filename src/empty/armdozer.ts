import type { Armdozer } from "../player/armdozer";
import type { Burst } from "../player/burst";
import type { ArmdozerState } from "../state/armdozer-state";

/** 空のバースト */
export const EMPTY_BURST: Burst = {
  type: "RecoverBattery",
  recoverBattery: 0,
  turnStartBatteryCorrect: 0,
};

/** 空のアームドーザデータ */
export const EMPTY_ARMDOZER: Armdozer = {
  id: "id",
  name: "name",
  maxHp: 3000,
  maxBattery: 5,
  power: 2000,
  speed: 2000,
  burst: EMPTY_BURST,
};

/** 空のアームドーザゲームステート */
export const EMPTY_ARMDOZER_STATE: ArmdozerState = {
  ...EMPTY_ARMDOZER,
  battery: EMPTY_ARMDOZER.maxBattery,
  hp: EMPTY_ARMDOZER.maxHp,
  effects: [],
  enableBurst: true,
};
