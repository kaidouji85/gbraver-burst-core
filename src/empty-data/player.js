// @flow

import type {PlayerState} from "../state/player-state";
import {EMPTY_ARMDOZER_STATE} from "./armdozer";
import {EMPTY_PILOT_STATE} from "./pilot";

/** 空のプレイヤーステート */
export const EMPTY_PLAYER_STATE: PlayerState = {
  playerId: '',
  armdozer: EMPTY_ARMDOZER_STATE,
  pilot: EMPTY_PILOT_STATE,
};
