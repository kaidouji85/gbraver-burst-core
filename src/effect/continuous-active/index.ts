import * as R from "ramda";

import { GameState, GameStateX } from "../../state/game-state";
import { removeBatteryRecoverSkip } from "../remove-battery-recover-skip";
import { removeTurnStartBatteryCorrect } from "../remove-turn-start-battery-correct";
import { TurnChange } from "../turn-change/turn-change";
import { hasContinuousActive } from "./has-continuous-active";
import { removeContinuousActive } from "./remove-continuous-active";

/**
 * アクティブプレイヤー継続を実行できるか否かを判定する
 * @param state ゲームステート
 * @returns 判定結果、trueで実行できる
 */
export function canContinuousActive(state: GameState): boolean {
  const { activePlayerId, players } = state;
  const activePlayer = players.find((p) => p.playerId === activePlayerId);
  return activePlayer
    ? hasContinuousActive(activePlayer.armdozer.effects)
    : false;
}

/**
 * アクティブプレイヤー継続を実行する
 * @param state 更新前のゲーム ステート
 * @returns 更新結果、実行不可能な場合はnullを返す
 */
export function continuousActive(state: GameState): GameStateX<TurnChange> {
  const { players, activePlayerId } = state;
  const activePlayer = players.find((p) => p.playerId === activePlayerId);
  if (!activePlayer) {
    throw new Error("not found active player");
  }

  const updatedActivePlayer = {
    ...activePlayer,
    armdozer: {
      ...activePlayer.armdozer,
      effects: R.pipe(
        removeContinuousActive,
        removeBatteryRecoverSkip,
        removeTurnStartBatteryCorrect,
      )(activePlayer.armdozer.effects),
    },
  };
  const updatedPlayers = players.map((p) =>
    p.playerId === activePlayer.playerId ? updatedActivePlayer : p,
  );
  const effect: TurnChange = {
    name: "TurnChange",
    recoverBattery: 0,
    reason: "ContinuousActive",
  };
  return { ...state, players: updatedPlayers, effect };
}
