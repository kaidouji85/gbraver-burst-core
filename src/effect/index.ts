import { z } from "zod";

import { BatteryDeclaration, BatteryDeclarationSchema } from "./battery-declaration/battery-declaration";
import { Battle, BattleSchema } from "./battle/battle";
import { BurstEffect, BurstEffectSchema } from "./burst/burst-effect";
import { GameEnd, GameEndSchema } from "./game-end/game-end";
import { InputCommand, InputCommandSchema } from "./input-command/input-command";
import { PilotSkillEffect, PilotSkillEffectSchema } from "./pilot-skill/pilot-skill-effect";
import { Reflect, ReflectSchema } from "./reflect/reflect";
import { RightItself, RightItselfSchema } from "./right-itself/right-itself";
import { StartGame, StartGameSchema } from "./start-game/start-game";
import { TurnChange, TurnChangeSchema } from "./turn-change/turn-change";
import { UpdateRemainingTurn, UpdateRemainingTurnSchema } from "./update-remaning-turn/update-remaining-turn";

/** ゲーム中に発生する効果をまとめたもの */
export type Effect =
  | StartGame
  | GameEnd
  | InputCommand
  | Reflect
  | BatteryDeclaration
  | Battle
  | TurnChange
  | BurstEffect
  | UpdateRemainingTurn
  | RightItself
  | PilotSkillEffect;

/** Effect zodスキーマ */
export const EffectSchema = z.union([
  StartGameSchema,
  GameEndSchema,
  InputCommandSchema,
  ReflectSchema,
  BatteryDeclarationSchema,
  BattleSchema,
  TurnChangeSchema,
  BurstEffectSchema,
  UpdateRemainingTurnSchema,
  RightItselfSchema,
  PilotSkillEffectSchema,
]);