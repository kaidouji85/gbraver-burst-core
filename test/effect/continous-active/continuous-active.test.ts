import {
  ArmdozerEffect,
  BatteryRecoverSkip,
  ContinuousActivePlayer,
  EMPTY_ARMDOZER_EFFECT,
  GameState,
  PlayerState,
  TurnStartBatteryCorrect,
} from "../../../src";
import { continuousActive } from "../../../src/effect/continuous-active";
import { EMPTY_ARMDOZER_STATE } from "../../../src/empty/armdozer";
import { EMPTY_GAME_STATE } from "../../../src/empty/game-state";
import { EMPTY_PLAYER_STATE } from "../../../src/empty/player";

/** 効果 アクティブプレイヤー継続 */
const CONTINUOUS_ACTIVE: ContinuousActivePlayer = {
  type: "ContinuousActivePlayer",
  period: {
    type: "SpecialPeriod",
  },
};

/** 効果 ターン開始時バッテリー回復スキップ */
const BATTERY_RECOVER_SKIP: BatteryRecoverSkip = {
  type: "BatteryRecoverSkip",
  period: {
    type: "SpecialPeriod",
  },
};

/** ターン開始時バッテリー回復量補正 */
const TURN_START_BATTERY_CORRECT: TurnStartBatteryCorrect = {
  type: "TurnStartBatteryCorrect",
  correctBattery: 1,
  period: {
    type: "SpecialPeriod",
  },
};

/**
 * 攻撃側プレイヤーを生成する
 * @param battery 現在のバッテリー値
 * @param effects 現在の効果
 * @returns 生成結果
 */
const createAttacker = (
  battery: number,
  effects: ArmdozerEffect[],
): PlayerState => ({
  ...EMPTY_PLAYER_STATE,
  playerId: "attacker",
  armdozer: {
    ...EMPTY_ARMDOZER_STATE,
    battery,
    maxBattery: 5,
    effects: [EMPTY_ARMDOZER_EFFECT, EMPTY_ARMDOZER_EFFECT, ...effects],
  },
});

/** 防御側プレイヤー */
const defender: PlayerState = {
  ...EMPTY_PLAYER_STATE,
  playerId: "defender",
  armdozer: {
    ...EMPTY_ARMDOZER_STATE,
    battery: 2,
    maxBattery: 5,
    effects: [CONTINUOUS_ACTIVE],
  },
};

test("アクティブプレイヤー継続が正しく処理できる", () => {
  const attacker = createAttacker(2, [CONTINUOUS_ACTIVE]);
  const lastState: GameState = {
    ...EMPTY_GAME_STATE,
    players: [defender, attacker],
    activePlayerId: attacker.playerId,
  };
  const result = continuousActive(lastState);
  expect(result).toMatchSnapshot("no-other-effects");
});

test("BatteryRecoverSkip、TurnStartBatteryCorrectは取り除かれる", () => {
  const attacker = createAttacker(2, [
    CONTINUOUS_ACTIVE,
    BATTERY_RECOVER_SKIP,
    TURN_START_BATTERY_CORRECT,
  ]);
  const lastState: GameState = {
    ...EMPTY_GAME_STATE,
    players: [defender, attacker],
    activePlayerId: attacker.playerId,
  };
  const result = continuousActive(lastState);
  expect(result).toMatchSnapshot("battery-recover-skip");
});
