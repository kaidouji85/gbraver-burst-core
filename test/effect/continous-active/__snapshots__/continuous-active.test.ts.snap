// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatteryRecoverSkip、TurnStartBatteryCorrectは取り除かれる 1`] = `
{
  "activePlayerId": "attacker",
  "effect": {
    "name": "TurnChange",
    "recoverBattery": 0,
  },
  "players": [
    {
      "armdozer": {
        "battery": 2,
        "burst": {
          "recoverBattery": 5,
          "turnStartBatteryCorrect": 1,
          "type": "RecoverBattery",
        },
        "effects": [
          {
            "period": {
              "type": "SpecialPeriod",
            },
            "type": "ContinuousActivePlayer",
          },
        ],
        "enableBurst": true,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "defender",
    },
    {
      "armdozer": {
        "battery": 2,
        "burst": {
          "recoverBattery": 5,
          "turnStartBatteryCorrect": 1,
          "type": "RecoverBattery",
        },
        "effects": [
          {
            "period": {
              "remainingTurn": 1,
              "type": "TurnLimit",
            },
            "type": "Empty",
          },
          {
            "period": {
              "remainingTurn": 1,
              "type": "TurnLimit",
            },
            "type": "Empty",
          },
        ],
        "enableBurst": true,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "attacker",
    },
  ],
}
`;

exports[`アクティブプレイヤー継続が正しく処理できる 1`] = `
{
  "activePlayerId": "attacker",
  "effect": {
    "name": "TurnChange",
    "recoverBattery": 0,
  },
  "players": [
    {
      "armdozer": {
        "battery": 2,
        "burst": {
          "recoverBattery": 5,
          "turnStartBatteryCorrect": 1,
          "type": "RecoverBattery",
        },
        "effects": [
          {
            "period": {
              "type": "SpecialPeriod",
            },
            "type": "ContinuousActivePlayer",
          },
        ],
        "enableBurst": true,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "defender",
    },
    {
      "armdozer": {
        "battery": 2,
        "burst": {
          "recoverBattery": 5,
          "turnStartBatteryCorrect": 1,
          "type": "RecoverBattery",
        },
        "effects": [
          {
            "period": {
              "remainingTurn": 1,
              "type": "TurnLimit",
            },
            "type": "Empty",
          },
          {
            "period": {
              "remainingTurn": 1,
              "type": "TurnLimit",
            },
            "type": "Empty",
          },
        ],
        "enableBurst": true,
        "hp": 3000,
        "id": "id",
        "maxBattery": 5,
        "maxHp": 3000,
        "name": "name",
        "power": 2000,
        "speed": 2000,
      },
      "pilot": {
        "enableSkill": true,
        "id": "EMPTY_PLAYER",
        "name": "名無し",
        "skill": {
          "recoverBattery": 2,
          "type": "RecoverBatterySkill",
        },
      },
      "playerId": "attacker",
    },
  ],
}
`;
